libs/components/packages/src/schematics/migrations/update-7/ag-grid-28.schematic.ts

```diff
@@ -0,0 +1,267 @@
+  agGridAllModulesEnt: NodeDependency | null
+): string {
+  // Replace `@ag-grid-community/all-modules` with `ag-grid-community`
+  if (agGridAllModules && updatedContent.includes(`${ANY_MODULE}all-modules`)) {
+    updatedContent = updatedContent.replace(
+      /@ag-grid-community\/all-modules/g,
+      AG_GRID
+    );
+  }
+
+  // Replace `@ag-grid-enterprise/all-modules` with `ag-grid-enterprise`
+  if (
+    agGridAllModulesEnt &&
+    updatedContent.includes(`${ENT_MODULE}all-modules`)
+  ) {
+    updatedContent = updatedContent.replace(
+      /@ag-grid-enterprise\/all-modules/g,
+      AG_GRID_ENT
+    );
+  }
+
+  // Replace `@ag-grid-community/angular` with `ag-grid-angular`
+  if (agGridAllModules && updatedContent.includes(`${ANY_MODULE}angular`)) {
+    updatedContent = updatedContent.replace(
+      /@ag-grid-community\/angular/g,
+      'ag-grid-angular'
+    );
+  }
+  return updatedContent;
+}
+
+/**
+ * Remove `AgGridModule.withComponents` in module imports.
+ */
+function removeWithComponentsStaticImportsCall(
+  filePath: Path,
+  updatedContent: string
+): string {
+  if (
+    (filePath.endsWith('module.ts') || filePath.endsWith('spec.ts')) &&
+    updatedContent.includes(withComponents)
+  ) {
+    let pos: number;
+    while ((pos = updatedContent.indexOf(withComponents)) !== -1) {
+      // Find closing parenthesis after `AgGridModule.withComponents`
+      const end = updatedContent.indexOf(')', pos);
+      // Leave `AgGridModule`, drop `.withComponents`.
+      updatedContent = [
+        updatedContent.substring(0, pos + 12),
+        updatedContent.substring(end + 1),
+      ].join('');
+    }
+  }
+  return updatedContent;
+}
+
+/**
+ * Column API renamed getSecondaryColumns / setSecondaryColumns to getSecondaryPivotColumns / setSecondaryPivotColumns.
+ */
+function renameColumnApiFunctionsInCode(updatedContent: string): string {
+  if (
+    updatedContent.includes(AG_GRID) &&
+    updatedContent.includes('etSecondaryColumns(')
+  ) {
+    updatedContent = updatedContent.replace(
+      /(?<=columnApi)\.(get|set)SecondaryColumns\(/g,
+      (_, x) => `.${x}PivotResultColumns(`
+    );
+  }
+  return updatedContent;
+}
+
+/**
+ * Visit all files and apply the changes.
+ */
+function updateSourceFiles(): Rule {
+  return (tree: Tree, context: SchematicContext) => {
+    const warned: string[] = [];
+
+    function warnOnce(message: string) {
+      if (!warned.includes(message)) {
+        warned.push(message);
+        context.logger.warn(message);
+      }
+    }
+
+    const {
+      agGridAllModules,
+      agGridDependency,
+      agGridAllModulesEnt,
+      agGridDependencyEnt,
+    } = readDependencies(tree);
+
+    // AG Grid is not installed, so we don't need to do anything.
+    if (
+      !agGridAllModules &&
+      !agGridDependency &&
+      !agGridAllModulesEnt &&
+      !agGridDependencyEnt
+    ) {
+      return;
+    }
+
+    ['src', 'projects'].forEach((sourceRoot) => {
```

Comment:
Each project's root directory can be configured in `angular.json`, so we should solidify the logic of this loop by reading directly from the workspace config.
```
const { workspace } = await getWorkspace(tree);

for (const project of workspace.projects.values()) {
  tree.getDir(project.root).visit((filePath: Path) => {
```

Reply:
Refactored to use `getWorkspace`.